In [1]:
import pandas as pd
import numpy as np

In [15]:
# this function will download the dataset into your local machine
import requests


def download(url, path):
    try:
        response = requests.get(url)
        response.raise_for_status()  # Check if the request was successful
        with open(path, 'wb') as file:
            file.write(response.content)
        print(f"File downloaded successfully and saved to {path}")
    except requests.exceptions.RequestException as e:
        print(f"Error downloading the file: {e}")



## Read Data
We use `pandas.read_csv()` function to read the csv file. In brackets, we put the file path along with a quataion mark so that pandas will read the file into a dataframe from that address. The file path can be either an URL or your local file address.  

Because the data does not include headers, we can add an argument `headers = None` inside the ` read_csv()` method so that pandas will not automatically set the first row as a header.  

You can also assign the dataset to any variable you create.

In [18]:
url = "https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBMDeveloperSkillsNetwork-DA0101EN-SkillsNetwork/labs/Data%20files/auto.csv"
path = "lab01.csv"

In [19]:
download(url, path)

File downloaded successfully and saved to lab01.csv


In [20]:
# Import pandas library
import pandas as pd

# Read the online file by the URL provides above, and assign it to variable "df"
df = pd.read_csv(path, header=None)

In [21]:
# show the first 5 rows using dataframe.head() method
print("The first 5 rows of the dataframe")
print(df.head(5))

The first 5 rows of the dataframe
   0    1            2    3    4     5            6    7      8     9   ...  \
0   3    ?  alfa-romero  gas  std   two  convertible  rwd  front  88.6  ...   
1   3    ?  alfa-romero  gas  std   two  convertible  rwd  front  88.6  ...   
2   1    ?  alfa-romero  gas  std   two    hatchback  rwd  front  94.5  ...   
3   2  164         audi  gas  std  four        sedan  fwd  front  99.8  ...   
4   2  164         audi  gas  std  four        sedan  4wd  front  99.4  ...   

    16    17    18    19    20   21    22  23  24     25  
0  130  mpfi  3.47  2.68   9.0  111  5000  21  27  13495  
1  130  mpfi  3.47  2.68   9.0  111  5000  21  27  16500  
2  152  mpfi  2.68  3.47   9.0  154  5000  19  26  16500  
3  109  mpfi  3.19  3.40  10.0  102  5500  24  30  13950  
4  136  mpfi  3.19  3.40   8.0  115  5500  18  22  17450  

[5 rows x 26 columns]


In [ ]:
# Check the bottom 10 rows of data frame "df".
print("The last 10 rows of the dataframe")
print(df.tail(10))

## Add Headers
Take a look at our dataset. Pandas automatically set the header with an integet starting from 0.  
To better describe our data, we can introduce a header. This information is available at the dataset's [homepage](https://archive.ics.uci.edu/ml/datasets/Car+Evaluation).
Thus, we have to add headers manually.  
First, we create a list "headers" that include all column names in order. Then, we use `dataframe.columns = headers` to replce the headers with the list we created

In [22]:
# create header list
headers = ["symboling","normalized-losses","make","fuel-type","aspiration", "num-of-doors","body-style",
         "drive-wheels","engine-location","wheel-base", "length","width","height","curb-weight","engine-type",
         "num-of-cylinders", "engine-size","fuel-system","bore","stroke","compression-ratio","horsepower",
         "peak-rpm","city-mpg","highway-mpg","price"]
print("headers\n", headers)

headers
 ['symboling', 'normalized-losses', 'make', 'fuel-type', 'aspiration', 'num-of-doors', 'body-style', 'drive-wheels', 'engine-location', 'wheel-base', 'length', 'width', 'height', 'curb-weight', 'engine-type', 'num-of-cylinders', 'engine-size', 'fuel-system', 'bore', 'stroke', 'compression-ratio', 'horsepower', 'peak-rpm', 'city-mpg', 'highway-mpg', 'price']


We replace headers and recheck our dataframe:

In [23]:
df.columns = headers
print(df.head(10))

   symboling normalized-losses         make fuel-type aspiration num-of-doors  \
0          3                 ?  alfa-romero       gas        std          two   
1          3                 ?  alfa-romero       gas        std          two   
2          1                 ?  alfa-romero       gas        std          two   
3          2               164         audi       gas        std         four   
4          2               164         audi       gas        std         four   
5          2                 ?         audi       gas        std          two   
6          1               158         audi       gas        std         four   
7          1                 ?         audi       gas        std         four   
8          1               158         audi       gas      turbo         four   
9          0                 ?         audi       gas      turbo          two   

    body-style drive-wheels engine-location  wheel-base  ...  engine-size  \
0  convertible          rwd    

We need to replace the "?" symbol with NaN so the dropna() can remove the missing values:

In [25]:
df1 = df.replace("?", np.nan)

We can drop missing values along the column "price" as follows:

In [26]:
df = df1.dropna(subset=["price"], axis=0)
print(df.head(20))

    symboling normalized-losses         make fuel-type aspiration  \
0           3               NaN  alfa-romero       gas        std   
1           3               NaN  alfa-romero       gas        std   
2           1               NaN  alfa-romero       gas        std   
3           2               164         audi       gas        std   
4           2               164         audi       gas        std   
5           2               NaN         audi       gas        std   
6           1               158         audi       gas        std   
7           1               NaN         audi       gas        std   
8           1               158         audi       gas      turbo   
10          2               192          bmw       gas        std   
11          0               192          bmw       gas        std   
12          0               188          bmw       gas        std   
13          0               188          bmw       gas        std   
14          1               NaN   

In [27]:
## Find the name of the columns of the dataframe
print(df.columns)

Index(['symboling', 'normalized-losses', 'make', 'fuel-type', 'aspiration',
       'num-of-doors', 'body-style', 'drive-wheels', 'engine-location',
       'wheel-base', 'length', 'width', 'height', 'curb-weight', 'engine-type',
       'num-of-cylinders', 'engine-size', 'fuel-system', 'bore', 'stroke',
       'compression-ratio', 'horsepower', 'peak-rpm', 'city-mpg',
       'highway-mpg', 'price'],
      dtype='object')


## Save Dataset
Correspondingly, Pandas enabled us to save the dataset to csv. By using the `dataframe.to_csv()` method, you can add the file path and name along with quaotation marks in the brackets.  

For example, if you would save the dataframe "df" as "automobile.csv" to your local machine, you may use the syntax below, where `index = False` means the row will not be written.  
`df.to_csv("automobile.csv", index = False)`  
We can also read and save other file formats. We can use similar functions like `dp.read_csv()` and `df.to_csv()` for other data formats. The functions are listed in the following table:  

## Read/Save Other Data Formats

| Data Format | Read Function      | Save Function      |
|-------------|---------------------|---------------------|
| csv         | `pd.read_csv()`     | `df.to_csv()`       |
| json        | `pd.read_json()`    | `df.to_json()`      |
| excel       | `pd.read_excel()`   | `df.to_excel()`     |
| hdf         | `pd.read_hdf()`     | `df.to_hdf()`       |
| sql         | `pd.read_sql()`     | `df.to_sql()`       |
| ...         | ...                 | ...                 |


## Basic Insight of Dataset
After reading data into Pandas dataframe, it is time for us to explore the dataset.  
There are several ways to obtain essential insights of the data to help us better understand our dataset.

## Data types
Data has a variety of types.  
The main types stored in Pandas dataframes are object, float, int, bool and datetime64. In order to better learn about each attribute, it is always good for us to know the data type of each column. In Pandas:

In [28]:
df.dtypes

symboling              int64
normalized-losses     object
make                  object
fuel-type             object
aspiration            object
num-of-doors          object
body-style            object
drive-wheels          object
engine-location       object
wheel-base           float64
length               float64
width                float64
height               float64
curb-weight            int64
engine-type           object
num-of-cylinders      object
engine-size            int64
fuel-system           object
bore                  object
stroke                object
compression-ratio    float64
horsepower            object
peak-rpm              object
city-mpg               int64
highway-mpg            int64
price                 object
dtype: object

a series with the data type of each column is returned

In [29]:
# Check the data type of data frame "df" by .dtypes
print(df.dtypes)

symboling              int64
normalized-losses     object
make                  object
fuel-type             object
aspiration            object
num-of-doors          object
body-style            object
drive-wheels          object
engine-location       object
wheel-base           float64
length               float64
width                float64
height               float64
curb-weight            int64
engine-type           object
num-of-cylinders      object
engine-size            int64
fuel-system           object
bore                  object
stroke                object
compression-ratio    float64
horsepower            object
peak-rpm              object
city-mpg               int64
highway-mpg            int64
price                 object
dtype: object


As shown above, it is clear to see that the data type of "symboling" and "curb-weight" are `int64`, "normalized-losses" is `object`, and "wheel-base" is `float64`, etc. These data types can be changed.

## Describe
If we would like to get a statistical summary of each column, such as count, column mean value, column standard deviation, etc. We use the `dataframe.describe()` method:

In [ ]:
dataframe.describe()

This method will provide various summary statistics, excluding `NaN` (Not a Number) values.

In [30]:
df.describe()

symboling  wheel-base      length       width      height  \
count  201.000000  201.000000  201.000000  201.000000  201.000000   
mean     0.840796   98.797015  174.200995   65.889055   53.766667   
std      1.254802    6.066366   12.322175    2.101471    2.447822   
min     -2.000000   86.600000  141.100000   60.300000   47.800000   
25%      0.000000   94.500000  166.800000   64.100000   52.000000   
50%      1.000000   97.000000  173.200000   65.500000   54.100000   
75%      2.000000  102.400000  183.500000   66.600000   55.500000   
max      3.000000  120.900000  208.100000   72.000000   59.800000   

       curb-weight  engine-size  compression-ratio    city-mpg  highway-mpg  
count   201.000000   201.000000         201.000000  201.000000   201.000000  
mean   2555.666667   126.875622          10.164279   25.179104    30.686567  
std     517.296727    41.546834           4.004965    6.423220     6.815150  
min    1488.000000    61.000000           7.000000   13.000000    16.000000  
25%    2169.000000    98.000000           8.600000   19.000000    25.000000  
50%    2414.000000   120.000000           9.000000   24.000000    30.000000  
75%    2926.000000   141.000000           9.400000   30.000000    34.000000  
max    4066.000000   326.000000          23.000000   49.000000    54.000000

This shows the statistical summary of all numeric-typed (int, float) columns.  
For example, the attribute "symboling" has 205 counts, the mean value of this column is 0.83, the standard deviation is 1.25, the minimum value is -2, 25th percentile is 0, 50th percentile is 1, 75th percentile is 2, and the maximum value is 3.  
However, what if we would also like to check all the columns including those that are of type object?  

You can add an argument `include = "all"` inside the bracket. Let's try it again.

In [31]:
# describe all the cloumns in "df"
df.describe(include = "all")

symboling normalized-losses    make fuel-type aspiration  \
count   201.000000               164     201       201        201   
unique         NaN                51      22         2          2   
top            NaN               161  toyota       gas        std   
freq           NaN                11      32       181        165   
mean      0.840796               NaN     NaN       NaN        NaN   
std       1.254802               NaN     NaN       NaN        NaN   
min      -2.000000               NaN     NaN       NaN        NaN   
25%       0.000000               NaN     NaN       NaN        NaN   
50%       1.000000               NaN     NaN       NaN        NaN   
75%       2.000000               NaN     NaN       NaN        NaN   
max       3.000000               NaN     NaN       NaN        NaN   

       num-of-doors body-style drive-wheels engine-location  wheel-base  ...  \
count           199        201          201             201  201.000000  ...   
unique            2          5            3               2         NaN  ...   
top            four      sedan          fwd           front         NaN  ...   
freq            113         94          118             198         NaN  ...   
mean            NaN        NaN          NaN             NaN   98.797015  ...   
std             NaN        NaN          NaN             NaN    6.066366  ...   
min             NaN        NaN          NaN             NaN   86.600000  ...   
25%             NaN        NaN          NaN             NaN   94.500000  ...   
50%             NaN        NaN          NaN             NaN   97.000000  ...   
75%             NaN        NaN          NaN             NaN  102.400000  ...   
max             NaN        NaN          NaN             NaN  120.900000  ...   

        engine-size  fuel-system  bore  stroke compression-ratio horsepower  \
count    201.000000          201   197     197        201.000000        199   
unique          NaN            8    38      36               NaN         58   
top             NaN         mpfi  3.62    3.40               NaN         68   
freq            NaN           92    23      19               NaN         19   
mean     126.875622          NaN   NaN     NaN         10.164279        NaN   
std       41.546834          NaN   NaN     NaN          4.004965        NaN   
min       61.000000          NaN   NaN     NaN          7.000000        NaN   
25%       98.000000          NaN   NaN     NaN          8.600000        NaN   
50%      120.000000          NaN   NaN     NaN          9.000000        NaN   
75%      141.000000          NaN   NaN     NaN          9.400000        NaN   
max      326.000000          NaN   NaN     NaN         23.000000        NaN   

        peak-rpm    city-mpg highway-mpg price  
count        199  201.000000  201.000000   201  
unique        22         NaN         NaN   186  
top         5500         NaN         NaN  8921  
freq          36         NaN         NaN     2  
mean         NaN   25.179104   30.686567   NaN  
std          NaN    6.423220    6.815150   NaN  
min          NaN   13.000000   16.000000   NaN  
25%          NaN   19.000000   25.000000   NaN  
50%          NaN   24.000000   30.000000   NaN  
75%          NaN   30.000000   34.000000   NaN  
max          NaN   49.000000   54.000000   NaN  

[11 rows x 26 columns]

Now it provides the statistical summary of all the columns, including object-typed attributes.  

We can now see how many unique values there are, which is the top value and the frequency of top value in the object-typed columns.  

Some values in the table above show as "NaN", this is because those numbers are not available regarding a particular column type.

In [ ]:
## apply the method to ".describe()" to the columns 'lenght' and 'compression-ratio'
df[['length', 'compression-ratio']].describe()